### ISSUE TITLE
`mapValues`: Narrow value types of mapped object

### ISSUE THREAD

**Issue by Robstei** [remeda/remeda#780]
Due to the way TypeScript works, there are some limitations ([1](https://github.com/microsoft/TypeScript/issues/31652),[2](https://stackoverflow.com/questions/55012174/why-doesnt-object-keys-return-a-keyof-type-in-typescript)) regarding functions that deal with objects and arrays.
The naive return type could be wrong hence the type is wider.

Is there a way in Remeda that allows mapping the values of an object (with as const) that preserves the exact key-value relationship?

```typescript

import * as R from 'remeda';

const ages = {
  marc: 22,
  jenny: 15,
} as const;

// Desired type can be created
type AgesAsStrings = { [Key in keyof typeof ages]: `${(typeof ages)[Key]}` };

// Object with desired type can be created manually
const agesAsStringsManually: AgesAsStrings = {
  marc: '22',
  jenny: '15',
};

// This works but the type AgesAsStrings is not assignable
const agesAsStringsRemeda = R.mapValues(ages, (value) => String(value));

console.log(agesAsStringsRemeda);

```

[Example in stackblitz](https://stackblitz.com/edit/vitejs-vite-rv3mnf?file=src%2Fmain.ts)

**Comment by cjquines** [comment]
with the example as-written, we can't do anything like this, because we don't have enough type information about the mapper. i would expect something like this to work though:

```ts
import * as R from 'remeda';

const ages = {
  marc: 22,
  jenny: 15,
} as const;

function mapper<T extends number>(value: T): `${T}` {
    return String(value) as `${T}`;
}

const agesAsStrings = R.mapValues(ages, mapper);
```

however, [see playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAKjgQwM5wEpwGZQiOAcigFMQSATZQgbgCg6BjCAO1XmQHMT0BeOAN504cEMiiMAXHABMMgDTC4AKxIsWAT2kBGAKyKAvinTM2MenWwBXFoxjBWo5GDAkoAHgAqcEgA8YahToLFYgAEZuAHwAFABuyAA2ViTSngCU0gAGACQCngaZgkoipDBWUCxwAMowUMAsnHGJyWnGcDl5BfQGDAD0vXCeABbA6ADu0ADW6GFW8DBDJHAwGq5wAILcqOuoNXUN6KNwLBAcqKjAnCzIYQkkTKzsKFs7e-WcfJgAdGJgAGrNHjRLg8eROFxuNL0foiEQAPQA-EA), this doesn't work; the returned type is `{ marc: "22" | "15"; jenny: "22" | "15"; }`.

looking at the type for `mapValues`, i don't see an easy fix for this. but i think it's a reasonable feature request.

**Comment by eranhirsch** [comment]
This isn't a problem just with mapValues, it exists in map too:

```ts
const data = [1, 2, 3] as const;
const mapped = map(data, x => x.toString());
//    ^? ["1" | "2" | "3", "1" | "2" | "3", "1" | "2" | "3"]
```

This is because when I call the mapping function, I already call it with the type `T[number]`, which takes all 3 values and puts them in the union: `1 | 2 | 3`.

There is no way for me to type the function itself differently per item, as its type is tied to the array itself and not to the invocation per item.

This is a holy-grail typing problem. I've only seen one project that said they solved it, and the typing was way too complex, so I couldn't extract something meaningful for use in Remeda.

**Comment by cjquines** [comment]
fair. this looks painful and is blocked on more TS features. i'll close as wontfix i think.
